File: backend/lib/viewmodels/meditation_viewmodel.dart
import 'dart:developer';

import 'package:flutter/material.dart';
import 'package:flutter/widgets.dart';
import 'package:just_audio/just_audio.dart';
import '../models/meditation_models.dart';
import '../services/meditation_repository.dart';
import '../services/monitoring_service.dart';
import 'package:flutter/foundation.dart' show kIsWeb;
import 'package:firebase_auth/firebase_auth.dart';
import 'package:cloud_firestore/cloud_firestore.dart';

class MeditationViewModel extends ChangeNotifier {
  final MeditationRepository _repository;
  final MonitoringService _monitoringService;
  final AudioPlayer _audioPlayer = AudioPlayer();

  List<MeditationSession> _sessions = [];
  List<MeditationSession> get sessions => List.unmodifiable(_sessions);
  MeditationSession? _currentSession;
  bool _isPlaying = false;
  double _currentProgress = 0.0;
  MeditationCategory? _selectedCategory;
  bool _audioInitialized = false;
  bool _isPremiumUser = false;

  MeditationSession? get currentSession => _currentSession;
  bool get isPlaying => _isPlaying;
  double get currentProgress => _currentProgress;
  MeditationCategory? get selectedCategory => _selectedCategory;
  bool get audioInitialized => _audioInitialized;
  bool get isWeb => kIsWeb;
  bool get isPremiumUser => _isPremiumUser;

  MeditationViewModel(this._repository, this._monitoringService) {
    _loadSessions();
    _setupAudioListener();
  }

  Future<void> _loadSessions() async {
    try {
      _sessions = await _repository.getAllSessions();
      
      // Optimización para web: precarga metadatos pero no audio
      if (kIsWeb) {
        for (var session in _sessions) {
          _preloadImageForWeb(session.imageUrl);
        }
      }
      
      notifyListeners();
    } catch (e) {
      log('Error cargando sesiones: $e');
    }
  }

  // Precarga imágenes para mejorar la experiencia en web
  void _preloadImageForWeb(String imageUrl) {
    if (kIsWeb) {
      final image = NetworkImage(imageUrl);
      precacheImage(image, null);
    }
  }

  void selectCategory(MeditationCategory? category) async {
    _selectedCategory = category;
    
    if (category != null) {
      try {
        _sessions = await _repository.getSessionsByCategory(category);
      } catch (e) {
        log('Error filtrando por categoría: $e');
      }
    } else {
      await _loadSessions();
    }
    
    notifyListeners();
  }

  Future<void> playSession(String sessionId) async {
    final session = _sessions.firstWhere((s) => s.id == sessionId);
    _currentSession = session;
    
    try {
      // Manejo especial para web
      if (kIsWeb) {
        // Inicializar audio solo cuando el usuario interactúa explícitamente
        await _audioPlayer.setUrl(session.audioUrl);
        _audioInitialized = true;
        
        // En web, esperamos a que el usuario presione play explícitamente
        // para cumplir con las políticas de autoplay
        _isPlaying = false;
      } else {
        await _audioPlayer.setUrl(session.audioUrl);
        await _audioPlayer.play();
        _isPlaying = true;
      }
      
      _monitoringService.logEvent('meditation_session_started', {
        'session_id': sessionId,
        'session_title': session.title,
        'session_category': session.category.toString(),
        'platform': kIsWeb ? 'web' : 'mobile',
      });
      
      notifyListeners();
    } catch (e) {
      log('Error reproduciendo sesión: $e');
    }
  }

  Future<void> resumeSession() async {
    try {
      await _audioPlayer.play();
      _isPlaying = true;
      notifyListeners();
    } catch (e) {
      log('Error al reanudar sesión: $e');
      // Manejo específico para errores web
      if (kIsWeb) {
        _showWebAudioError();
      }
    }
  }

  void pauseSession() {
    _audioPlayer.pause();
    _isPlaying = false;
    notifyListeners();
  }

  Future<void> stopSession() async {
    if (_currentSession != null) {
      _monitoringService.logEvent('meditation_session_completed', {
        'session_id': _currentSession!.id,
        'session_title': _currentSession!.title,
        'session_duration': _currentProgress * _currentSession!.duration,
        'platform': kIsWeb ? 'web' : 'mobile',
      });
    }
    
    await _audioPlayer.stop();
    _currentSession = null;
    _isPlaying = false;
    _currentProgress = 0.0;
    _audioInitialized = false;
    notifyListeners();
  }

  void _setupAudioListener() {
    _audioPlayer.positionStream.listen((position) {
      if (_currentSession != null) {
        _currentProgress = position.inSeconds / _currentSession!.duration;
        
        if (_currentProgress >= 0.99) {
          stopSession();
        }
        
        notifyListeners();
      }
    });

    // Manejo de errores específico para web
    if (kIsWeb) {
      _audioPlayer.playerStateStream.listen((state) {
        if (state.processingState == ProcessingState.completed) {
          stopSession();
        }
      });
    }
  }

  void _showWebAudioError() {
    // Esta función sería implementada en la UI para mostrar un mensaje
    // informando al usuario sobre problemas de reproducción en web
    log('Error de reproducción en navegador web');
  }

  @override
  void dispose() {
    _audioPlayer.dispose();
    super.dispose();
  }

  Future<void> checkSubscriptionStatus() async {
    final user = FirebaseAuth.instance.currentUser;
    if (user != null) {
      final subscriptions = await FirebaseFirestore.instance
          .collection('users')
          .doc(user.uid)
          .collection('subscriptions')
          .where('isActive', isEqualTo: true)
          .where('endDate', isGreaterThan: DateTime.now().toIso8601String())
          .get();

      _isPremiumUser = subscriptions.docs.isNotEmpty;
      notifyListeners();
    }
  }

  Future<List<MeditationSession>> getAllSessions() async {
    final sessions = await _repository.getAllSessions();
    if (!_isPremiumUser) {
      return sessions.where((session) => !session.isPremium).toList();
    }
    return sessions;
  }
}
 

File: backend/feature/meditation/src/main/java/com/mindcare/meditation/MeditationViewModel.kt
package com.mindcare.meditation

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.mindcare.analytics.AnalyticsEvent
import com.mindcare.analytics.AnalyticsManager
import com.mindcare.meditation.data.repository.MeditationRepository
import com.mindcare.meditation.model.MeditationCategory
import com.mindcare.meditation.model.MeditationProgress
import com.mindcare.meditation.model.MeditationSession
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.*
import kotlinx.coroutines.launch
import javax.inject.Inject

@HiltViewModel
class MeditationViewModel @Inject constructor(
    private val meditationRepository: MeditationRepository,
    private val analyticsManager: AnalyticsManager
) : ViewModel() {

    private val _sessions = MutableStateFlow<List<MeditationSession>>(emptyList())
    val sessions = _sessions.asStateFlow()

    private val _currentSession = MutableStateFlow<MeditationSession?>(null)
    val currentSession = _currentSession.asStateFlow()

    private val _isPlaying = MutableStateFlow(false)
    val isPlaying = _isPlaying.asStateFlow()

    private val _currentProgress = MutableStateFlow(0f)
    val currentProgress = _currentProgress.asStateFlow()

    private val _selectedCategory = MutableStateFlow<MeditationCategory?>(null)
    val selectedCategory = _selectedCategory.asStateFlow()

    init {
        loadSessions()
        observeAudioProgress()
    }

    private fun loadSessions() {
        viewModelScope.launch {
            meditationRepository.getAllSessions()
                .catch { error ->
                    // Manejar error
                }
                .collect { sessions ->
                    _sessions.value = sessions
                }
        }
    }

    fun selectCategory(category: MeditationCategory?) {
        _selectedCategory.value = category
        if (category != null) {
            viewModelScope.launch {
                meditationRepository.getSessionsByCategory(category)
                    .collect { sessions ->
                        _sessions.value = sessions
                    }
            }
        } else {
            loadSessions()
        }
    }

    fun playSession(sessionId: String) {
        viewModelScope.launch {
            val session = sessions.value.find { it.id == sessionId }
            session?.let {
                _currentSession.value = it
                meditationRepository.startSession(sessionId)
                _isPlaying.value = true
                analyticsManager.logEvent(AnalyticsEvent.MeditationSessionStarted(sessionId))
            }
        }
    }

    fun pauseSession() {
        viewModelScope.launch {
            meditationRepository.pauseSession()
            _isPlaying.value = false
        }
    }

    fun resumeSession() {
        viewModelScope.launch {
            meditationRepository.resumeSession()
            _isPlaying.value = true
        }
    }

    fun stopSession() {
        viewModelScope.launch {
            currentSession.value?.let { session ->
                meditationRepository.stopSession(session.id)
                analyticsManager.logEvent(AnalyticsEvent.MeditationSessionCompleted(session.id))
            }
            _currentSession.value = null
            _isPlaying.value = false
        }
    }

    private fun observeAudioProgress() {
        viewModelScope.launch {
            meditationRepository.currentAudioProgress
                .collect { progress ->
                    _currentProgress.value = progress
                    if (progress >= 0.99f) {
                        stopSession()
                    }
                }
        }
    }

    override fun onCleared() {
        super.onCleared()
        viewModelScope.launch {
            meditationRepository.pauseSession()
        }
    }
} 